design-proposals/cluster-orch-integrating-k8s-into-emt.md

```diff
@@ -0,0 +1,253 @@
+**2. Manually through a direct request to the Cluster Manager**: This method allows for more customized cluster configurations and management.
+
+#### Automatic Cluster Creation
+
+The first option introduces a streamlined process for users seeking simplicity. Both the Web UI and CLI will offer a toggle for `Cluster Join Automatically`. When enabled, users can choose to either select an existing cluster to join or create a new one. By default, the system will create a new single-node K3s cluster with default baseline template for the host, requiring no additional inputs from the user.
+
+The sequence diagram below illustrates the API call flow for host registration when `Cluster Join Automatically` is enabled.
+
+```mermaid
+sequenceDiagram
+    actor User
+    User->>+UI: Register a new host
+    opt
+    UI->>+Cluster Manager: Create a cluster
+    Cluster Manager-->>-UI: Success
+    end
+    UI->>+Infra Manager: Register host
+    Infra Manager-->>-UI: Success
+    UI->>+Infra Manager: Create instance
+    Infra Manager-->>-UI: Success
+    UI->>+Cluster Manager: Add host to the cluster
+    Cluster Manager-->>-UI: Success
+    UI-->>-User: Success
+```
+
+*[todo: discuss transaction management and failure handling]*
+*[todo: update sequence diagram with failure scenario]*
+
+A new cluster is created only when the user does not provide an existing cluster name and opts for the default action, which is creating a new single-node cluster. In the final step, the host is added to either the user-provided existing cluster or the newly created cluster.
+
+*[todo: add another sequence diagram for host connect, cluster reconciliation pause/unpause]*
+
+#### Manual Cluster Creation
+
+Manual cluster creation is an existing method where users create a cluster by selecting a template and hosts. Before we talk about manual cluster creation workflow, it is important to understand the coupling between EMT image version and cluster template introduced by integrating Kubernetes into EMT. For instance, creating a K3s cluster with a K3s v1.32 template may not work as expected on a EMT machine provisioned with K3s v1.30. Such discrepancies can occur when Edge Orchestration undergoes multiple upgrades, resulting in a mix of multiple versions of cluster templates and EMT machines. While this issue is relevant for K3s on EMT only, given that this combination is the primary use case, it is important to address it properly in the user workflow. There are a few options to mitigate this issue.
+
+**Option 1)** Restrict cluster creation to onboarded hosts only, excluding provisioned hosts. The Cluster Manager automatically selects the appropriate EMT version, and requests instance creation to the Infra Manager to install the OS and Kubernetes. Could be simplest option.
```

Comment:
doable by looking at the content of the manifest.json associated with an OSProfile. It tells basically all the package versions

Reply:
Thanks! This is the part that I wanted confirmation.

Reply:
Who would be doing that? UI? Cluster Orch? Some component must parse the manifest and infer the installed k3s version. Please clarify in the proposal